File: 1155-number-of-dice-rolls-with-target-sum/1155-number-of-dice-rolls-with-target-sum.cpp
class Solution {
public:
    int numRollsToTarget(int n, int k, int target) {
        const int MOD = 1000000007;

        if (n * k < target) {
            return 0;
        }

        vector<vector<long long>> dp(n + 1, vector<long long>(target + 1, 0));

        dp[0][0] = 1;

        for (int i = 1; i <= n; i++) {
            for (int j = i; j <= min(i * k, target); j++) {
                for (int t = 1; t <= min(k, j); t++) {
                    dp[i][j] = (dp[i][j] + dp[i - 1][j - t]) % MOD;
                }
            }
        }

        return static_cast<int>(dp[n][target]);
    }
};

File: 1155-number-of-dice-rolls-with-target-sum/1155-number-of-dice-rolls-with-target-sum.go
const mod = 1000000007

func numRollsToTarget(n, k, target int) int {
    dp := make([][]int, n+1)
    for i := range dp {
        dp[i] = make([]int, target+1)
    }

    dp[0][0] = 1

    for i := 1; i <= n; i++ {
        for j := 1; j <= target; j++ {
            for t := 1; t <= k; t++ {
                if j-t >= 0 {
                    dp[i][j] = (dp[i][j] + dp[i-1][j-t]) % mod
                }
            }
        }
    }

    return dp[n][target]
}
